src/server/serveractiveobject.cpp

```diff
@@ -73,3 +73,21 @@ void ServerActiveObject::dumpAOMessagesToQueue(std::queue<ActiveObjectMessage> &
 		m_messages_out.pop();
 	}
 }
+
+void ServerActiveObject::markForRemoval()
+{
+	if (!m_pending_removal) {
+		if (!isGone())
+			onMarkedForRemoval();
+		m_pending_removal = true;
+	}
+}
+
+void ServerActiveObject::markForDeactivation()
+{
+	if (!m_pending_deactivation) {
```

Comment:
Is there a articular reason for this check? (same for code above)
`if (!isGone())` already does this, and a value assignment isn't expensive.

Reply:
Okay, so:
- There are two _competing_ deactivation paths, and possible interactions between them are pretty murky, this whole system isn't designed right. Lifecycles should probably be nested, not side by side.
- I needed to make sure the deactivation api is never ever called twice. This call checks whether the other path has been already invoked, without attempting to figure out if it ever could actually happen.
- Using `isGone` instead of the opposite path looks a bit cleaner, and since `isGone` is inline, the compiler ought to chip away the redundant part.

Basically, it was either this or reworking activation entirely and possibly making this too complex for a timely merge.

Reply:
Thought of a better way to do this, the redundant check now happens elsewhere but makes a **lot** more sense and also acts as a sanity check and documents something. (on_deactivate is guaranteed to always have a valid object state)